docs/reference/troubleshooting/corruption-issues.asciidoc

```diff
@@ -0,0 +1,82 @@
+[[corruption-troubleshooting]]
+== Troubleshooting corruption
+
+{es} expects that the data it reads from disk is exactly the data it previously
+wrote. If it detects that the data on disk is different from what it wrote then
+it will report some kind of exception such as:
+
+- `org.apache.lucene.index.CorruptIndexException`
+- `org.elasticsearch.gateway.CorruptStateException`
+- `org.elasticsearch.index.translog.TranslogCorruptedException`
+
+Typically these exceptions happen due to a checksum mismatch. Most of the data
+that {es} writes to disk is followed by a checksum using a simple algorithm
+known as CRC32 which is fast to compute and good at detecting the kinds of
+random corruption that may happen when using faulty storage. A CRC32 checksum
+mismatch definitely indicates that something is faulty, although of course a
+matching checksum doesn't prove the absence of corruption.
+
+Verifying a checksum is expensive since it involves reading every byte of the
+file which takes significant effort and might evict more useful data from the
+filesystem cache, so systems typically doesn't verify the checksum on a file
+very often. This is why you tend only to encounter a corruption exception when
+something unusual is happening. For instance, corruptions are often detected
+during merges, shard movements, and snapshots. This does not mean that these
+proceses are causing corruption: they are examples of the rare times where
+reading a whole file is necessary. {es} takes the opportunity to verify the
+checksum at the same time, and this is when the corruption is detected and
+reported. It doesn't indicate the cause of the corruption or when it happened.
+Corruptions can remain undetected for many months.
+
+The files that make up a Lucene index are written sequentially from start to
+end and then never modified or overwritten. This access pattern means the
+checksum computation is very simple and can happen on-the-fly as the file is
+written, and also makes it very unlikely that an incorrect checksum is due to a
```

Comment:
I think I agree to this. But we do not calculate it on every write, like when transferring the file over network or from snapshot. I'd also think the risk of a user space bug in those transfers is small and the argument as a whole stands.

The only change I would request here is to add "initially", i.e.:

```suggestion
checksum computation is very simple and can happen on-the-fly as the file is initially
written, and also makes it very unlikely that an incorrect checksum is due to a
```

Reply:
We do verify checksums when copying files over the network during a recovery. You're right that we don't when restoring from a snapshot, but I think we should (I opened https://github.com/elastic/elasticsearch/issues/88878). I'll commit this change.